src/content/blog/first-four-val-town-runtimes.mdx

```diff
@@ -0,0 +1,247 @@
+farther away from environment variables or anything sensitive.
+
+All in all, this was an improvement, but we still had further to go.
+
+- Security problems were far fewer and had much less impact. Deno's sandboxing
+  was a huge unlock. However, it was still possible to escape the sandbox, by
+  doing things like spawning a worker inside of the worker. We were also still
+  trying to preserve [a complex security model](https://blog.val.town/blog/restricted-library-mode/)
+  in which JavaScript functions had their own environment variables and
+  security isolation.
+
+<figure>
+  <Image src={memoryImage} alt="Our server restarting" />
+  <figcaption>
+    Our server repeatedly restarting because of memory issues
+  </figcaption>
+</figure>
+
+- However, Workers triggered [memory problems](https://github.com/denoland/deno/issues/18414)
+  that made the evaluation server slower and slower over time until it restarted.
+  We were spawning and killing a _lot_ of workers, and understandably this
+  wasn't a typical usecase.
+- Server isolation had a performance cost: for a val that powers a website, for each
+  request we'd receive the request in the Express server, serialize it, send
+  it to Deno, which would deserialize and operate on it, send the response to Express,
+  which would then send it to the user.
+
+Despite these issues, the switch to Deno was a big upgrade, and we were significantly
+less affected by anything that untrusted code could do.
+
+#### Runtime 4: Node.js + Deno processes
+
+The current iteration of our runtime learns from the lessons so far:
+
+- You can't really use JavaScript to build a JavaScript sandbox.
+- Deno is a powerful tool for simple containment of functionality.
+- Memory leaks are scary.
+
+Right now, Vals run as temporary Deno subprocesses of a Node.js server.
+The subprocess's code is tiny: it uses [node-deno-vm](https://github.com/casual-simulation/node-deno-vm),
+a terrific module from Kallyn Gowdy, which works by exposing a WebSocket
+connection between Deno and Node.js and wrapping it in
+a [postMessage](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage)-style API.
+The "guest" code that wraps vals is also now [public](https://esm.town/guest.ts),
+though it's still an implementation detail and could change at any time.
+
+While this gives far better isolation between user code and system code
+than our previous vm and Worker-based strategies, it takes a little more
+time to start up, so we use a pool of pre-initialized processes to
+avoid that time penalty.
+
+This strategy also gives the ability to have more fine-grained communication
+with Deno, because it's based on WebSockets instead of a simple REST
+HTTP request cycle. We can send multiple messages to control vals while
+they're running, and receive multiple messages from them as they run.
+This is key to how we just cut 100ms from typical val runtimes - more on
+that later.
```

Comment:
> more on that later.

This made me think we'd talk about it later in this post. I'd be more explicit that this will be a future post.

```suggestion
This is key to how we just cut 100ms from typical val runtimes - more on
that on a future post.
```